jni/src/faiss_wrapper.cpp

```diff
@@ -0,0 +1,357 @@
+void TrainIndex(faiss::Index * index, faiss::Index::idx_t n, const float* x);
+
+void knn_jni::faiss_wrapper::createIndex(JNIEnv * env, jintArray idsJ, jobjectArray vectorsJ, jstring indexPathJ,
+                                         jobject parametersJ) {
+
+    try {
+        if (idsJ == nullptr) {
+            throw std::runtime_error("IDs cannot be null");
+        }
+
+        if (vectorsJ == nullptr) {
+            throw std::runtime_error("Vectors cannot be null");
+        }
+
+        if (indexPathJ == nullptr) {
+            throw std::runtime_error("Index path cannot be null");
+        }
+
+        if (parametersJ == nullptr) {
+            throw std::runtime_error("Parameters cannot be null");
+        }
+    } catch (...) {
+        CatchCppExceptionAndThrowJava(env);
+    }
+
+    // parametersJ is a Java Map<String, Object>. ConvertJavaMapToCppMap converts it to a c++ map<string, jobject>
+    // so that it is easier to access.
+    auto parametersCpp = knn_jni::ConvertJavaMapToCppMap(env, parametersJ);
+
+    // Get space type for this index
+    if(parametersCpp.find(knn_jni::SPACE_TYPE) == parametersCpp.end()) {
+        throw std::runtime_error("Space type not found");
+    }
+    jobject spaceTypeJ = parametersCpp[knn_jni::SPACE_TYPE];
+    std::string spaceTypeCpp(knn_jni::ConvertJavaObjectToCppString(env, spaceTypeJ));
+    faiss::MetricType metric = TranslateSpaceToMetric(spaceTypeCpp);
+
+    // Read data set
+    int numVectors = knn_jni::GetJavaObjectArrayLength(env, vectorsJ);
+    int numIds = knn_jni::GetJavaIntArrayLength(env, idsJ);
+    if (numIds != numVectors) {
+        throw std::runtime_error("Number of IDs does not match number of vectors");
+    }
+
+    int dim = knn_jni::GetInnerDimensionOf2dJavaArray(env, vectorsJ);
+    auto dataset = knn_jni::Convert2dJavaObjectArrayToCppFloatVector(env, vectorsJ, dim);
+
+    // Create faiss index
+    if(parametersCpp.find(knn_jni::METHOD) == parametersCpp.end()) {
+        throw std::runtime_error("No method passed");
+    }
+    jobject indexDescriptionJ = parametersCpp[knn_jni::METHOD];
+    std::string indexDescriptionCpp(knn_jni::ConvertJavaObjectToCppString(env, indexDescriptionJ));
+
+    std::unique_ptr<faiss::Index> indexWriter;
+    indexWriter.reset(faiss::index_factory(dim, indexDescriptionCpp.c_str(), metric));
+
+    // Add extra parameters that cant be configured with the index factory
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        jobject subParametersJ = parametersCpp[knn_jni::PARAMETERS];
+        auto subParametersCpp = knn_jni::ConvertJavaMapToCppMap(env, subParametersJ);
+        SetExtraParameters(env, subParametersCpp, indexWriter.get());
+        env->DeleteLocalRef(subParametersJ);
+    }
+
+    // Train index if needed -- check if there is a case where index needs part trained but is trained
+    if(!indexWriter->is_trained) {
+        //TODO: What needs to be freed???
```

Comment:
Everything allocated should be wrapped in an auto_ptr or similar smart pointer construct to do things like `env->DeleteLocalRef`. I'd ensure this before this is merged.

Reply:
Makes sense. I will ensure all native allocations are wrapped in smart pointers. For calls to free references in jenv, I will work on some kind of wrapper to ensure everything is freed.

Reply:
So, DeleteLocalRef is more of an optimization. Local references will automatically be freed when jni call returns.